### ISSUE TITLE
Random "syntax error: operand expected" errors

### ISSUE THREAD

**Issue by rux616** [akinomyoga/ble.sh#92]
**ble version**: 0.4.0-devel3+054e5c1
**Bash version**: 5.0.17(1)-release, x86_64-pc-linux-gnu

Over the past few days, ble.sh has been breaking (seemingly randomly) and spitting out this error message:

```
bash: ((: '0': syntax error: operand expected (error token is "'0'")
bash: '0': syntax error: operand expected (error token is "'0'")
```

Sometimes the `0` is a different number, but mostly it's `0`.

I'll describe my setup to see if you need more information:

* I use Ubuntu 20.04 on Windows WSL2 and also Ubuntu 20.04 on a VM I use for work remotely, both have seen the errors
* I use Alacritty v0.7.1 as my terminal emulator
* On my remote VM I use tmux and tend to have something like 30 panes spread out over various windows and sessions
* My bash dotfiles can be found [here](https://github.com/rux616/init/tree/master/bash)

Once the break happens, I need to hard-kill the parent bash process in some manner, using tmux's `kill-pane` command if I'm in tmux, but sometimes having to outright close the Alacritty window. Once the break happens, the following keys produce the error message above:

* arrow keys
* `ins`, `home`, `del`, `end` keys above the arrow keys
* `enter` key
* `tab` key
* `backspace` key
* numbers (symbols e.g. `#` and `&` are fine)
* `-` key (`_` is fine)
* numpad number keys
* numpad `enter` key
* numpad `-` key

I went on vacation on Feb 12, and didn't get back until about Wednesday this week (Feb 24), so I'm thinking the offending change happened sometime during that time (or maybe a little before, as I was winding down for a couple days to go on said vacation), as ble.sh was working fine before.

Let me know if I can provide any other information for you.

**Comment by akinomyoga** [comment]
Thank you for the report. OK, I have recently pushed large changes related to #82, so I think now `ble.sh` is a bit unstable. So far I haven't met the problem, but I'll later try your configuration to see if it reproduces them in my environment.

### Can you try an older version of ble.sh?

> I went on vacation on Feb 12, and didn't get back until about Wednesday this week (Feb 24), so I'm thinking the offending change happened sometime during that time (or maybe a little before, as I was winding down for a couple days to go on said vacation), as ble.sh was working fine before.

For the time being, can you check out the commit 3cae6e4 (which is the last commit before the large change) and use it for a while to confirm that the problem doesn't occur in that version?

```bash
$ cd /path/to/bash/repository
$ git checkout 3cae6e4
```

When you finish the testing, you can check out the `master` branch.

```bash
$ git checkout master
```

It depends on how frequent the error message occurs, but, if it is easy to check if a particular version has a problem or not, you may also try other commits between 3cae6e4...054e5c1.

### Questions on keys

> Once the break happens, the following keys produce the error message above:

Maybe the problem is related to the user-input decoder. I have now checked the recent change on the decoder quickly, but have no idea until now.

- Q. Can I assume normal alphabet keys don't produce the error message? If the error message is not produced, does they function as usual (i.e., is the input letter inserted into the command-line text)?
- Q. How about <kbd>C-l</kbd> or other Ctrl + alphabets? Do they function as usual? I guess the answer is no because <kbd>RET</kbd> (<kbd>C-m</kbd>) seems not working, but let me confirm it.
- Q. Also, how many error messages are repeated for one keystroke of <kbd>home</kbd>? Can I assume that just one pair of the reported messages is produced on a keystroke (see the below example)?

```
# For example, something like this?
bash: ((: '0': syntax error: operand expected (error token is "'0'")
bash: '0': syntax error: operand expected (error token is "'0'")

# Or, like this for one <home>?
bash: ((: '0': syntax error: operand expected (error token is "'0'")
bash: '0': syntax error: operand expected (error token is "'0'")
bash: ((: '0': syntax error: operand expected (error token is "'0'")
bash: '0': syntax error: operand expected (error token is "'0'")
bash: ((: '0': syntax error: operand expected (error token is "'0'")
bash: '0': syntax error: operand expected (error token is "'0'")
bash: ((: '0': syntax error: operand expected (error token is "'0'")
bash: '0': syntax error: operand expected (error token is "'0'")
```

**Comment by rux616** [comment]
> For the time being, can you check out the commit [3cae6e4](https://github.com/akinomyoga/ble.sh/commit/3cae6e4dc525633401a357e71150f8237ababe81) (which is the last commit before the large change) and use it for a while to confirm that the problem doesn't occur in that version?

Sure, I'll use that one for now.

> It depends on how frequent the error message occurs, but, if it is easy to check if a particular version has a problem or not, you may also try other commits between [3cae6e4...054e5c1](https://github.com/akinomyoga/ble.sh/compare/3cae6e4dc525633401a357e71150f8237ababe81...054e5c1b5ccf0a5aeed613bf907c83f20f32269c).

If I get the chance, I will try to narrow down in exactly which commit the instability started.

> Q. Can I assume normal alphabet keys don't produce the error message? If the error message is not produced, does they function as usual (i.e., is the input letter inserted into the command-line text)?

Yes, that is correct. Additionally, any text that was there appears again and is appended to.

> Q. How about C-l or other Ctrl + alphabets? Do they function as usual? I guess the answer is no because RET (C-m) seems not working, but let me confirm it.

Most `CTRL+<alpha>` combinations  don't work, but here are the ones that do:

* `C-q`
* `C-d`
* `C-l`
* `C-x`
* `C-v`

> Q. Also, how many error messages are repeated for one keystroke of home? Can I assume that just one pair of the reported messages is produced on a keystroke (see the below example)?

This one:

> ```
> # For example, something like this?
> bash: ((: '0': syntax error: operand expected (error token is "'0'")
> bash: '0': syntax error: operand expected (error token is "'0'")
> ```

---

I have been experimenting a bit more (still running 0.4.0-devel3+054e5c1), and I think I might have found a trigger on my system (maybe _the_ trigger, but I can't be sure):

* (start bash normally, with `source /path/to/ble.sh` and eventually `[[ ${BLE_VERSION-} ]] && ble-attach`)
* run the command `echo foo`
* hit the 'up' arrow to get the last run command from history
* hit the 'down' arrow to get back to a clean command line
* hit the 'down' arrow again
* 💥  watch the errors flow

**Comment by akinomyoga** [comment]
Ah, OK! I could reproduce it in my environment with your `init.sh`! I think I can fix it soon.

**Comment by akinomyoga** [comment]
I pushed a fix e9be69e. The problem was introduced in the change 5f9adfe to fix the problem reported at https://github.com/akinomyoga/ble.sh/issues/85#issuecomment-782765297. Thank you for the report!

**Comment by rux616** [comment]
I forgot to respond saying that it's all fixed. Thanks for the quick work!
